bridges/snowbridge/primitives/inbound-queue/src/v2/converter.rs

```diff
@@ -89,19 +104,79 @@ where
 		fn prepare(message: Message) -> Result<PreparedMessage, ConvertMessageError> {
 			let mut remote_xcm: Xcm<()> = Xcm::new();
 
-			// Allow xcm decode failure so that assets can be trapped on AH instead of this
-			// message failing but funds are already locked on Ethereum.
-			if let Ok(versioned_xcm) = VersionedXcm::<()>::decode_with_depth_limit(
-				MAX_XCM_DECODE_DEPTH,
-				&mut message.xcm.as_ref(),
-			) {
-				if let Ok(decoded_xcm) = versioned_xcm.try_into() {
-					remote_xcm = decoded_xcm;
-				}
-			}
-
 			let ether_location = Location::new(2, [GlobalConsensus(EthereumNetwork::get())]);
 
+			match &message.xcm {
+				XcmCommand::Raw(raw) => {
+					// Allow xcm decode failure so that assets can be trapped on AH instead of this
+					// message failing but funds are already locked on Ethereum.
+					if let Ok(versioned_xcm) = VersionedXcm::<()>::decode_with_depth_limit(
+						MAX_XCM_DECODE_DEPTH,
+						&mut raw.as_ref(),
+					) {
+						if let Ok(decoded_xcm) = versioned_xcm.try_into() {
+							remote_xcm = decoded_xcm;
+						}
+					}
+				},
+				XcmCommand::TokenRegistration { token, network } => {
+					let chain_id = match EthereumNetwork::get() {
+						Ethereum { chain_id } => chain_id,
+						_ => return Err(ConvertMessageError::InvalidNetwork),
+					};
+					let bridge_owner =
+						EthereumLocationsConverterFor::<[u8; 32]>::from_chain_id(&chain_id);
+					let dot_asset = Location::new(1, Here);
+					let dot_fee: xcm::prelude::Asset = (dot_asset, CreateAssetDeposit::get()).into();
+
+					let asset_deposit: xcm::prelude::Asset =
+						(ether_location.clone(), message.value).into();
+
+					let create_call_index: [u8; 2] = CreateAssetCall::get();
+					let asset_id = Location::new(
+						2,
+						[GlobalConsensus(EthereumNetwork::get()), AccountKey20 { network: None, key: (*token).into() }],
+					);
+
+					match *network {
+						// Polkadot
+						0 => {
+							let instructions = vec![
+								// Exchange eth for dot to pay the asset creation deposit
+								ExchangeAsset {
+									give: asset_deposit.clone().into(),
+									want: dot_fee.clone().into(),
+									maximal: false,
+								},
+								// Deposit the dot deposit into the bridge sovereign account (where the asset
+								// creation fee will be deducted from)
+								DepositAsset { assets: dot_fee.into(), beneficiary: bridge_owner.into() },
+								// Call to create the asset.
+								Transact {
+									origin_kind: OriginKind::Xcm,
+									fallback_max_weight: None,
+									call: (
+										create_call_index,
+										asset_id,
+										MultiAddress::<[u8; 32], ()>::Id(bridge_owner.into()),
+										MINIMUM_DEPOSIT,
+									)
+										.encode()
+										.into(),
+								},
+								ExpectTransactStatus(MaybeErrorCode::Success),
+							];
+							remote_xcm = instructions.into();
+						},
+						// Kusama
+						1 => {
+							// TODO
```

Comment:
Lets rather tackle this as part of the separate ticket I created: https://linear.app/snowfork/issue/SNO-1371. As I want to merge this soon so Parity can review.

For, now just raise an `InvalidNetwork` error.

Reply:
Sure, https://github.com/vgeddes/polkadot-sdk/pull/13/commits/194ab7c9a4b932525b28975610a8a53d3100f06c.